numpyro/handlers.py

```diff
@@ -242,6 +243,66 @@ def process_message(self, msg):
             msg['stop'] = True
 
 
+class collapse(trace):
+    """
+    EXPERIMENTAL Collapses all sites in the context by lazily sampling and
+    attempting to use conjugacy relations. If no conjugacy is known this will
+    fail. Code using the results of sample sites must be written to accept
+    Funsors rather than Tensors. This requires ``funsor`` to be installed.
+    """
+    def __init__(self, *args, **kwargs):
+        import funsor
+
+        funsor.set_backend("jax")
+        super().__init__(*args, **kwargs)
+
+    def process_message(self, msg):
+        from funsor.terms import Funsor
+
+        if msg["type"] == "sample":
+            if msg["value"] is None:
+                msg["value"] = msg["name"]
```

Comment:
Is there no need for `msg["done"] = True` as in Pyro?

Reply:
We don't have `msg["done"]` in NumPyro. What does it do?

Reply:
@eb8680 can you explain `msg["done"]` vs `msg["stop"]`?

Reply:
`msg["done"]` indicates that `msg["value"]` is final and should not be changed by subsequent handlers.  In this case it would equivalent to `msg["value"] is not None`, although it can occasionally be useful to distinguish between these situations.